### PR TITLE
AAR-123 + AAR-131: unit tests for refactorings and `if-else` constructs

### PR DESCRIPTION
**Pull request by voddan**
Added tests for `if-else`;
Fixed the algorithm to react to the `else` branch

### CODE REVIEW

**Comment by voddan** [comment]
Depends on [PR-1](https://github.com/MergeTool/MergeTool/pull/2)

File: merge/choice.py [hunk 1]
```diff
@@ -2,10 +2,10 @@
 
 
 class Choice(Enum):
-    undesided = 0
+    undecided = 0
```
Comment (avriy) [inline merge/choice.py:5]: maybe `resolved`?

File: merge/conflict.py [hunk 2]
```diff
@@ -87,6 +148,14 @@ def description(self) -> str:
         return "\n left = \n" + self.left + "\n right = \n" + self.right
 
 
+"""
+A type alias for exporting `Conflict3Way` and `Conflict2Way` into other modules.
```
Comment (avriy) [inline merge/conflict.py:152]: why indention here is different from the one that is on lines from 125 to 130

File: merge/external_parser_setup.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+
+class ExternalParserSetup:
+    is_set = False
+
+    @classmethod
+    def setup(cls):
+        if not cls.is_set:
+            cls.is_set = True
+            from clang.cindex import Config
+            Config.set_library_file("/usr/local/opt/llvm/lib/libclang.dylib")
```
Comment (avriy) [inline merge/external_parser_setup.py:10]: don't like hardcoding(

File: merge/project_merge.py [hunk 4]
```diff
@@ -66,6 +66,7 @@ def compile_print(self):
     def parse(path: Path, tmp_path: Path):  # -> ProjectMerge:
         merges = []
         for file in path.iterdir():
-            merges.append(FileMerge.parse(file))
+            stream = file.open('r', encoding="latin-1")
```
Comment (avriy) [inline merge/project_merge.py]: why not utf-8?

File: test/merge/test_file_merge.py [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+
+
+class TestFileMerge(TestCase):
+    @classmethod
+    def setUpClass(cls):
+        from merge.external_parser_setup import ExternalParserSetup
+        ExternalParserSetup.setup()
+
+    def setUp(self):
+        self.fb1 = FileBit(1, ("int main() {\n"
```
Comment (avriy) [inline test/merge/test_file_merge.py:22]: maybe we can have these merge conflict examples in a files?

File: test/merge/test_file_merge.py [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+
+        parsed_file = FileMerge.parse(Path("./testproject/prog.cpp"), StringIO(text))
+
+        self.assertEqual(parsed_file.path, self.file_merge.path)
+        self.assertListEqual(parsed_file.file_bits, self.file_merge.file_bits)
+        self.assertListEqual(parsed_file.conflicts, self.file_merge.conflicts)
+
+    def test_can_parse(self):
+        self.assertTrue(FileMerge.can_parse(Path("./testproject/prog.cpp")))
+        self.assertTrue(FileMerge.can_parse(Path("/Users/voddan/MergeTool/prototype/testproject/prog.hpp")))
```
Comment (avriy) [inline test/merge/test_file_merge.py]: this hardcoding is very bad(

**Review by avriy — COMMENTED** [review]
needs work

**Comment by voddan** [comment]
@avriy 
Thanks for commenting!

Altho I would prefer that we went through the PRs in order, PR-1 then PR-2. 
Let's keep the order in the future ;)

File: merge/choice.py (hunk 1 above)
Comment (voddan) [reply merge/choice.py:5]: No, a resolved conflict is ether `left`, `right` or `both` (see the `is_resolved` method)

`Choice.undecided` is the opposite of  a resolved conflict

File: merge/conflict.py (hunk 2 above)
Comment (voddan) [reply merge/conflict.py:152]: Because it is a package level alias (almost like in Swift:)

File: test/merge/test_file_merge.py (hunk 5 above)
Comment (voddan) [reply test/merge/test_file_merge.py:22]: In what sense "in files"? Please elaborate

**Comment by avriy** [comment]
Your PR-2 was from the branch that was branched from the one in PR-1, so PR-2 has everything. Next time please create separate requests only for not dependant changes

File: test/merge/test_file_merge.py (hunk 5 above)
Comment (avriy) [reply test/merge/test_file_merge.py:22]: we've decided to keep merge conflict examples in a separate files

**Comment by voddan** [comment]
The changes were dependent, but related to different issues. That's why I thought that creating two PRs would be appropriate. 

My plan was to merge PR-1 with corrections, then update PR-2 for the new `master`,  then discuss PR-2, which by that time would only contain the relevant commits (since the master will have moved)

**Comment by avriy** [comment]
if it is a case, you should then open 2nd pull request only after 1st one is merged

File: test/merge/test_file_merge.py (hunk 5 above)
Comment (voddan) [reply test/merge/test_file_merge.py:22]: I still don't understand. 

We have merge examples on Confluence, but they do not relate to code.

Also we have a test project (./testproject) and a couple of sample files (./samples/*) for testing the whole program together. They are where they were, and I run the code on them too.

On the other hand, the tests in this file are "unit-test", since they test only one unit - the merge algorithm. They do not involve the CUI and other helper facilities.

To sum it up, we have global tests for the whole project, and we now have unit tests for separate parts. There is no reason to make the unit tests more complicated than they are.

**Comment by voddan** [comment]
Fair enough, I don't know why I didn't think of that myself :/

File: merge/external_parser_setup.py (hunk 3 above)
Comment (voddan) [reply merge/external_parser_setup.py:10]: Me neither, but this is something we have already had before. Presumably it is the default path to the library. 

The only way to remove this hard code is to create a search for this binary, cash it somewhere, then use it. This was planed for the future as a part of making the whole product cross platform and easily deployable.

Summing up, fixing it is complicated and deserves its own PR. I propose we leave this hard coded for now and create an issue to fix it later.

**Comment by voddan** [comment]
Implemented the relevant suggestions

**Comment by voddan** [comment]
So ... do we merge it?
